feat: Support Importing Precomputed Embeddings

Ayushjhawar8 opened the pull request:
/claim #149 
closes #149

Ayushjhawar8 commented:
Hey @ChuckHend ! I'm first time contributor to tembo. Please can you review my implementation to resolve the issue.

ChuckHend commented:
Hey @Ayushjhawar8, that you for the PR. The title of this PR says "Add document chunking capability to vectorize.table()" but the changes and bounty look like more closely related to "Bring your own embeddings?"

Ayushjhawar8 commented:
Hi @ChuckHend. Sry for bad pr title..i changed it....also im giving my idea behind the changes...so I implemented the `import_embeddings` feature to let users bring their pre-computed embeddings directly into pg_vectorize. I designed it to work with both join and append table methods since I wanted to maintain compatibility with existing workflows. The core idea was to avoid forcing users to recompute embeddings when they already have them - instead, I added validation to ensure the dimensions match what's expected, handled transaction safety to prevent partial imports, and made sure to clean up any pending realtime jobs. I built this specifically thinking about users who might be using Sentence-Transformers or similar models and want to leverage our management capabilities without redoing their embedding work. I also made sure to document it clearly in the API docs and added comprehensive tests to verify it works as intended across different scenarios. 
Please review the changes and tell me if it was helpful :)

Ayushjhawar8 commented:
Hi @ChuckHend. Sry for bad pr title..i changed it....also im giving my idea behind the changes...so i implemented the import_embeddings feature to let users bring their pre-computed embeddings directly into pg_vectorize. I designed it to work with both join and append table methods since I wanted to maintain compatibility with existing workflows. I built this specifically thinking about users who might be using Sentence-Transformers or similar models and want to take the management capabilities without redoing their embedding work. I also made sure to document it clearly in the API docs and added comprehensive tests to verify it works as intended across different scenarios.
Please review the changes and tell me if it was helpful :)

ChuckHend commented on docs/api/index.md:
````diff
@@ -1,3 +1,52 @@
+
+```sql
+-- First create a vectorize project
+SELECT vectorize.table(
+    job_name => 'product_search',
+    table => 'products',
+    primary_key => 'id',
+    columns => ARRAY['description'],
+    transformer => 'sentence-transformers/all-MiniLM-L6-v2'
+);
````
How does this interact with tthe `schedule` parameter on table()? I think when the user has it set to "realtime", then it will immediately start to generate embeddings for all records. I think there could be conflicts when import_embedding is called?

The default for `schedule` is a cron job, so I think it will work so long as the user calls import_embeddings() before the cron job kicks in. But that might get messy.

Ayushjhawar8 replied:
You raise a good point about potential conflicts with the schedule parameter. Here's how i visioned it should works:

1. For "realtime" schedule:
- The `import_embeddings` function already handles this case. It includes cleanup logic that deletes any pending jobs from the queue for that project:
```rust
if job_params.schedule == "realtime" {
    let delete_q = format!(
        "DELETE FROM pgmq.q_{} WHERE message->>'job_name' = $1",
        VECTORIZE_QUEUE
    );
    Spi::run_with_args(
        &delete_q,
        Some(vec![(PgBuiltInOids::TEXTOID.oid(), job_name.into_datum())]),
    )?;
}
```
This ensures no conflicts between imported embeddings and auto-generated ones.

2. For cron schedule (default):
- You're right that timing could be tricky, but the upsert logic in `import_embeddings` uses ON CONFLICT DO UPDATE, so even if a cron job runs, it won't overwrite imported embeddings unless they're actually different.
- i think user should:
  a) Import embeddings immediately after creating the table, before the first cron run
  b) Or temporarily disable the cron job, import embeddings, then re-enable

Should i make the docs more clearer for users in regards to this?

Ayushjhawar8 replied:
Yeah, that makes sense. The "realtime" schedule kicking in right away does make it tricky to import embeddings without conflicts. Maybe we could:

- Add a defer_embedding option to vectorize.table(), so embeddings aren’t generated immediately.
- Introduce a "manual" schedule option alongside 'realtime' and cron, where embeddings are only created when explicitly triggered.
- Adjust import_embeddings to temporarily pause embedding generation during import.

I’m leaning toward the "manual" schedule option since it keeps things simple. What do you think?

ChuckHend replied:
what do you think about creating a new function, something like (maybe there is a better name) `table_from()` that would take care of first getting the embeddings in place, prior to creating the triggers or scheduling the cron job?

Ayushjhawar8 replied:
Thats a clean solution i think. Having `vectorize.table_from()` would give us a dedicated path for pre-loaded embeddings that takes existing embeddings as input, handles the full setup in one call, sets up triggers/cron only after embeddings are in place
Okay i would like to explore this approach.

Ayushjhawar8 replied:
Hey @ChuckHend i tried to implement the `vectorize.table_from()` function....please can you check.

ChuckHend replied:
Looks pretty good! I'll try it out soon and let you know if I run into anything!

Ayushjhawar8 commented:
Hey @ChuckHend i see many test failing can you help me resolve it, i will also try to fix the ones i could scratch my head around

ChuckHend commented:
Don't worry about any tests failing because of missing API keys or permissions, such as the two below.

```
    test_cohere
    test_private_hf_model
```

This one and other though I think needs a fix?
https://github.com/tembo-io/pg_vectorize/actions/runs/13410514940/job/37651761054?pr=199

Ayushjhawar8 commented:
okay I will be working to fix that

Ayushjhawar8 commented:
hey @ChuckHend please can you check i tried to fix some

ChuckHend commented on extension/src/api.rs:
```diff
@@ -285,3 +289,177 @@ fn chunk_text(document: &str, max_characters: i64) -> Vec<String> {
+                job_name, job_params.primary_key, job_params.primary_key
+            );
+
+            Spi::run_with_args(
+                &insert_q,
+                Some(vec![
+                    (PgBuiltInOids::TEXTOID.oid(), original_id.into_datum()),
+                    (PgBuiltInOids::FLOAT8ARRAYOID.oid(), embeddings.into_datum()),
+                ]),
+            )?;
```
Although, I think this could get REALLY slow if there are a lot of embeddings that already exist. Instead, maybe this could be crafted into a select statement, something like:

```sql
INSERT INTO vectorize._embeddings_{}
SELECT {}, {} FROM {};
```

Of course, filling in the variables correctly.

Ayushjhawar8 replied:
i tried to implement this, please tell if approach is right?

ChuckHend commented on extension/src/api.rs:
```diff
@@ -285,3 +289,177 @@ fn chunk_text(document: &str, max_characters: i64) -> Vec<String> {
+        job_name,
+        schema,
+        table,
+        columns,
```
For now I think it will be fine to just clone here. That will help get your tests to pass :) 

```suggestion
        columns.clone(),
```

ChuckHend commented:
@Ayushjhawar8 , I left a few comments. This is coming along nicely!

Ayushjhawar8 commented:
Hey @ChuckHend , I tried to implement all your suggestion as:
1. For the borrow issue, i tried to now clone the columns vector when passing it to `init_table`
2. For the `Spi::select` issue, I've completely rewritten the approach to use bulk SQL operations

Hoping it fixes the test, Also please review it further if needed. 

> @Ayushjhawar8 , I left a few comments. This is coming along nicely!
Thanks :)

Ayushjhawar8 commented:
umm why install dependency one is failing now

Also its not format issue now ? 😃

ChuckHend commented:
Looks like some things in our CI environment need to be updated. I'll fix that today.

ChuckHend commented on extension/src/api.rs:
```diff
@@ -285,3 +289,162 @@ fn chunk_text(document: &str, max_characters: i64) -> Vec<String> {
+        .build()
+        .unwrap_or_else(|e| error!("failed to initialize tokio runtime: {}", e));
+
+    let expected_dim =
+        runtime.block_on(async { provider.model_dim(&meta.transformer.api_name()).await })?;
+
+    let mut count = 0;
```
these two variables are unused. You can catch this by run `cargo clippy` at from the `./extension` directory.

Ayushjhawar8 replied:
Um count is actually being used in several places in the code so we should not remove it, i removed the expected_dim as it was unused as per our new approach

ChuckHend replied:
Ah, you're right. For some reason the compiler doesn't register that the `count` it used in that `format!()` call.

Ayushjhawar8 replied:
Ya i think its being overwritten before being used that's why its not registering it

Ayushjhawar8 commented:
hm still that test failing let me see how can i work around it

Ayushjhawar8 commented:
is there a way i can test integration test in my local every time?

ChuckHend commented:
@Ayushjhawar8 , I pushed a few changes in https://github.com/tembo-io/pg_vectorize/pull/199/commits/53c2aceea232125bcafd4896d499887afbc4dfc1
- removed unused variables
- refactored handled of that `count` variable so compiler would know how we're using it
- added quotes around the "table" parameter in vectorize.table(). This is just a bad choice on the API design on our part. We should rename that so that you dont have to put double quotes around it.

ChuckHend reviewed (commented):
I think all that remains for a test to cover `vectorize.table_from()`? Looks like import_embeddings() is functioning as intended now.

Ayushjhawar8 commented:
> @Ayushjhawar8 , I pushed a few changes in [53c2ace](https://github.com/tembo-io/pg_vectorize/commit/53c2aceea232125bcafd4896d499887afbc4dfc1)
> 
> * removed unused variables
> * refactored handled of that `count` variable so compiler would know how we're using it
> * added quotes around the "table" parameter in vectorize.table(). This was a bad choice on the API design on our part, but we'll be correcting it soon. Sorry for having to deal with that.

wow thanks for the improvements! That makes things much clearer. Appreciate the help! 💖

Ayushjhawar8 commented:
> I think all that remains for a test to cover `vectorize.table_from()`? Looks like import_embeddings() is functioning as intended now.

im on it 🫡

Ayushjhawar8 commented:
oh table one issue again, i need to put quotes around it ig

ChuckHend commented:
> I'm not sure what is this error about? We are using `realtime` schedule though?

I think its happening during the [init_table()](https://github.com/tembo-io/pg_vectorize/blob/6c13bf576b36d05237c1ffd352107c9dd7392be2/extension/src/api.rs#L403) call earlier in from_table(). I think init_table() will need to be modified such that it can accept a value of "manual" so that it will not init the cron job or table triggers.

Ayushjhawar8 commented on extension/tests/integration_tests.rs:
```diff
@@ -1381,7 +1381,7 @@ async fn test_table_from() {
             src_embeddings_col => 'embeddings',
             transformer => 'sentence-transformers/all-MiniLM-L6-v2',
-            schedule => 'realtime'
+            schedule => 'manual'
```
also i think we are already handling this in `api.rs` as 

In the `table_from` function in `src/api.rs`. 

1.  the initialization to use "manual" schedule:
```rust
init_table(
    // ... other params ...
    "manual", // Changed from schedule to "manual" to prevent immediate job creation
)?;
```

2. scheduling setup after importing embeddings:
```rust 
// Now set up the triggers or cron job based on the desired schedule
if schedule == "realtime" {
    // Create triggers for realtime updates
    let trigger_handler = create_trigger_handler(job_name, &columns, primary_key);
    Spi::run(&trigger_handler)?;

    let insert_trigger = create_event_trigger(job_name, schema, table, "INSERT");
    let update_trigger = create_event_trigger(job_name, schema, table, "UPDATE");

    Spi::run(&insert_trigger)?;
    Spi::run(&update_trigger)?;
} else if schedule != "manual" {
    // Set up cron job for scheduled updates
    init_cron(schedule, job_name)?;
}
```
do we need to make more changes in this?

ChuckHend replied:
I'll have to dive a bit deeper into the error and see exactly where it is coming from.

ChuckHend replied:
I think the error you are seeing is due to a bug unrelated to this feature. I'll see if I can fix it in https://github.com/tembo-io/pg_vectorize/pull/206

ChuckHend replied:
I think tests pass now!

ChuckHend reviewed (approved):
awesome, nice work, thank you!
